File: integration/DovahLinkValidationClient/CapabilitiesPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// A <c>capabilities</c> message payload (<c>protocol/schema/README.md</c>'s <c>capabilities</c>),
/// sent by both endpoints after <c>hello_ack</c>. No capability is currently registered, so both
/// the Bridge's and this client's own list are always empty.
/// </summary>
/// <param name="Capabilities">The capabilities advertised or requested by the sender.</param>
public sealed record CapabilitiesPayload(IReadOnlyList<Capability> Capabilities)
{
    /// <summary>
    /// Encodes this payload as a JSON object.
    /// </summary>
    /// <returns>The encoded <c>capabilities</c> payload.</returns>
    public JsonObject Encode()
    {
        var array = new JsonArray();
        foreach (Capability capability in Capabilities)
        {
            array.Add(capability.Encode());
        }
        return new JsonObject { ["capabilities"] = array };
    }

    /// <summary>
    /// Decodes and validates one <c>capabilities</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded capabilities payload.</returns>
    /// <exception cref="FormatException">Thrown when <c>capabilities</c> is missing, not an array,
    /// or contains a malformed entry.</exception>
    public static CapabilitiesPayload Decode(JsonObject payload)
    {
        JsonArray array = payload["capabilities"] as JsonArray
            ?? throw new FormatException("Missing or malformed capabilities.");
        var capabilities = new List<Capability>();
        foreach (JsonNode? node in array)
        {
            JsonObject entry = node as JsonObject ?? throw new FormatException("Capability entry must be an object.");
            capabilities.Add(Capability.Decode(entry));
        }
        return new CapabilitiesPayload(capabilities);
    }
}

File: sdk/dart/dovahlink_client/lib/src/protocol/capabilities_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/capability.dart';
import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';

part 'capabilities_payload.g.dart';

/// A `capabilities` message payload (`protocol/schema/README.md`'s `capabilities`), sent by both
/// endpoints after `hello_ack`. No capability is currently registered, so both the Bridge's and
/// this SDK's own list are always empty.
@JsonSerializable(checked: true, explicitToJson: true)
class CapabilitiesPayload {
  /// The capabilities advertised or requested by the sender.
  @JsonKey(required: true)
  final List<Capability> capabilities;

  /// Creates a capabilities payload.
  const CapabilitiesPayload({required this.capabilities});

  /// Decodes and validates one capabilities payload.
  factory CapabilitiesPayload.fromJson(JsonMap json) {
    try {
      return _$CapabilitiesPayloadFromJson(json);
    } on Object catch (error) {
      throw ProtocolFormatException('Invalid capabilities payload: $error');
    }
  }

  /// Encodes this payload as a JSON object.
  JsonMap toJson() => _$CapabilitiesPayloadToJson(this);
}
